src/Features/CSharp/Portable/SignatureHelp/AbstractCSharpSignatureHelpProvider.cs

```diff
@@ -72,5 +100,308 @@ protected IList<TaggedText> GetAwaitableUsage(IMethodSymbol method, SemanticMode
+                    if (foundParam >= 0)
+                    {
+                        found = foundParam;
+                    }
+                    else
+                    {
+                        var firstUnspecified = FirstUnspecifiedParameter();
+                        found = firstUnspecified < 0 ? 0 : firstUnspecified;
+                    }
+                }
+                else
+                {
+                    found = argumentIndexToSave >= 0 ? argToParamMap[argumentIndexToSave] : -1;
+                }
+
+                Debug.Assert(found < parameterCount);
+
+                return true;
+            }
+
+            // Find the parameter index corresponding to each argument provided 
+            bool PrepareArgToParamMap()
+            {
+                var currentParameterIndex = 0;
+                var seenOutOfPositionArgument = false;
+                var inParams = false;
+
+                for (var argumentIndex = 0; argumentIndex < argumentCount; argumentIndex++)
+                {
+                    if (argumentIndex >= parameterCount && !inParams)
+                    {
+                        return false;
+                    }
+
+                    var argument = arguments[argumentIndex];
+                    if (HasName(argument, out var name))
+                    {
+                        var namedParameterIndex = parameters.IndexOf(p => p.Name == name);
+                        if (namedParameterIndex < 0)
+                        {
+                            return false;
+                        }
+
+                        if (namedParameterIndex != currentParameterIndex)
+                        {
+                            seenOutOfPositionArgument = true;
+                        }
+
+                        AddArgToParamMapping(argumentIndex, namedParameterIndex);
+                        IncrementParameterIndexIfNeeded();
+                    }
+                    else if (IsEmptyArgument(argument.Expression))
+                    {
+                        if (!seenOutOfPositionArgument)
+                        {
+                            // We count the empty argument as a used position
+                            AddArgToParamMapping(argumentIndex, currentParameterIndex);
+                            IncrementParameterIndexIfNeeded();
+                        }
+                    }
+                    else if (seenOutOfPositionArgument)
+                    {
+                        // Unnamed arguments are not allowed after an out-of-position argument
+                        return false;
+                    }
+                    else
+                    {
+                        AddArgToParamMapping(argumentIndex, currentParameterIndex);
+                        IncrementParameterIndexIfNeeded();
+                    }
+                }
+
+                return true;
+
+                void IncrementParameterIndexIfNeeded()
```

Comment:
we're now double nested in.  i have to admit i've lost track of how this function works.  could you consider just extracting out a simple helper struct with some obvious state and simpler sibling methods?